{% extends 'main/base.html' %}
{% load static %}
{% block content %}


<div class="jumbotron">
  <h4 class="display-4">What is imputation?</h4>
  <h5 class="display-5">Try to solve the puzzle by filling in the missing letters in the sentence below.</h5>
  <p></p>
  <h5 class="text-success">Wh<font color="#ee0000">_</font> is my g<font color="#ee0000">_</font>nome inco<font color="#ee0000">_</font>plete?</h5>
  <p></p>
    <button type="button" class="btn btn-outline-primary btn-sm" data-toggle="modal" data-target="#Mymodal">Show Answer</button>
    <div class="modal" id="Mymodal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true" data-backdrop="false">
      <div class="modal-dialog" role="document">
        <div class="modal-content">
          <div class="modal-header">
            <h5 class="modal-title">Answer!</h5>
            <button type="button" class="close" data-dismiss="modal" aria-label="Close">
              <span aria-hidden="true">&times;</span>
            </button>
          </div>
          <div class="modal-body">
            <p class="text-success">Wh<font color="#ee0000">_</font> is my g<font color="#ee0000">_</font>nome inco<font color="#ee0000">_</font>plete?</p>
            <p class="text-info">Why is my genome incomplete?</p>
          </div>
          <div class="modal-footer">
            <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
          </div>
        </div>
      </div>
    </div>
  <br></br><p></p>
  <p>If you were able to complete the puzzle, you can understand how genotype imputation works. Genotype imputation is performed using a similar method to the one your brain used to fill in the missing letters above.
  You used context clues from the letters and words surrounding the missing letters as well as your knowledge of the known words in the english dictionary.
  Genotype imputation relies on a reference population of many individuals who have essentially no missing letters and looks for shared <a href="https://en.wikipedia.org/wiki/Haplotype" target="_blank">haplotypes <i class="fa fa-external-link"></i></a> between reference sequences and your sequence.
  Imputation algorithms provide a probability for each possible genotype at every imputed locus in the genome.</p>
</div>

<div class="jumbotron">
  <h5 class="display-5">References</h5>
  <p>The Imputation pipeline was architected with <a href="http://pgxcentre.github.io/genipe/" target="_blank">genipe <i class="fa fa-external-link"></i></a><br>
    Within genipe:<br>
    <a href="https://www.cog-genomics.org/plink2" target="_blank">plink <i class="fa fa-external-link"></i></a> is used for pre-phasing data cleaning and format conversion.<br>
    <a href="http://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html" target="_blank">SHAPEIT <i class="fa fa-external-link"></i></a> is used for phasing.<br>
    <a href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#home" target="_blank">IMPUTE2 <i class="fa fa-external-link"></i></a> is used for imputation, using the 1000 Genomes Phase 3 reference data.
  </p>
</div>

{% endblock %}
